Fix bootstrapping with Xcode command line tools

adamjstewart opened the pull request:
Apparently #31792 only fixed bootstrapping when using the full Xcode suite, this PR fixes bootstrapping when using only the Xcode command line tools.

Fixes #31889

adamjstewart committed:
python: simplify `libs()` and `headers()` methods in `python/package.py`

- [x] Rework `headers` to search a sequence of directories and to display all searched
      locations on error, as opposed to handling each directory with a variable

- [x] Make `headers` and `libs` do the same type of search and raise the same sort of
      errors.

Co-authored-by: Todd Gamblin <tgamblin@llnl.gov>

adamjstewart commented:
@rountree can you try the latest commit?

rountree commented:
@adamjstewart 

```
[+] /Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz
==> Installing clingo-bootstrap-spack-j4orhw72ixaxxks6y7dibclwsbqafxyr
==> No binary for clingo-bootstrap-spack-j4orhw72ixaxxks6y7dibclwsbqafxyr found: installing from source
==> Error: TypeError: expected str, bytes or os.PathLike object, not generator

/Users/rountree4/spack/var/spack/repos/builtin/packages/python/package.py:1116, in headers:
       1113                directories.append(os.path.join(self.config_vars["base"], subdir))
       1114            directories = dedupe(directories)
       1115            for directory in directories:
  >>   1116                headers = find_headers("pyconfig", directories)
       1117                if headers:
       1118                    config_h = headers[0]
       1119                    break
```

adamjstewart commented:
It's been a long day 😆

rountree commented:
No worries at all.  Rebuilding....

rountree commented:
Getting longer.... want me to create a new issue or do you think it's related?

```
==> No binary for clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz found: installing from source
==> Using cached archive: /Users/rountree4/spack/var/spack/cache/_source-cache/git//potassco/clingo.git/2a025667090d71b2c9dce60fe924feb6bde8f667.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Already patched clingo-bootstrap
==> clingo-bootstrap: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz/lib;/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz/lib64;/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib' '-DCMAKE_PREFIX_PATH:STRING=/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/re2c-2.2-ax6z5gjajrwybsvlnktzrohv2qqhkzbn;/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz;/Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/bison-3.8.2-m6v6mtbqsy7ws5ydetr7za6fewxnhw5l;/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8' '-DCLINGO_BUILD_WITH_LUA=OFF' '-DCLINGO_REQUIRE_PYTHON=ON' '-DCLINGO_BUILD_WITH_PYTHON=ON' '-DPYCLINGO_USER_INSTALL=OFF' '-DPYCLINGO_USE_INSTALL_PREFIX=ON' '-DCLINGO_BUILD_PY_SHARED:STRING=OFF' '-DPython_EXECUTABLE:STRING=/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/bin/python3.8' '-DPython_INCLUDE_DIR:STRING=/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/include/python3.8' '-DCLINGO_BUILD_APPS:STRING=OFF' '/var/folders/0p/2_2pgh5x52z6pl6xxlt2cyv0001l2r/T/rountree4/spack-stage/spack-stage-clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz/spack-src'

1 error found in build log:
  >> 3    CMake Error at /Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandle
          StandardArgs.cmake:230 (message):
     4      Could NOT find Python (missing: Python_LIBRARIES Interpreter Development
     5      Development.Module Development.Embed) (found version "3.8.9")
     6    Call Stack (most recent call first):
     7      /Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.
          cmake:594 (_FPHSA_FAILURE_MESSAGE)
     8      /Users/rountree4/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPython.cmake:561 (find_pac
          kage_handle_standard_args)
     9      CMakeLists.txt:130 (find_package)
```

rountree commented:
Nevermind, it's #31735 

So yes, this PR works well enough to get me to the next bug.  Thanks!

adamjstewart commented:
#31735 should already be fixed in develop, if you're still seeing issues then that's definitely related. Where does `libpython3.8.dylib` or `libpython3.8.a` live on your system?

adamjstewart commented:
Could also be that `libs` works correctly and we just need to pass it to the clingo build, let me try that.

adamjstewart commented:
Try the latest commit. If that raises an error saying it can't find the library, let me know where those libraries can be found. Probably going to be in that same root directory.

rountree commented:
```
% find /Library -name libpython3.8.dylib
...
/Library/Application Support/Adobe/Reader/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/config-3.8-darwin/libpython3.8.dylib
/Library/Application Support/Adobe/Reader/Frameworks/Python3.framework/Versions/3.8/lib/libpython3.8.dylib
...
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/config-3.8-darwin/libpython3.8.dylib
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/libpython3.8.dylib
```

```
% find /Library -name libpython3.8.a
...
/Library/Application Support/Adobe/Reader/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/config-3.8-darwin/libpython3.8.a
...
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/config-3.8-darwin/libpython3.8.a
```

rountree commented:
Testing the latest....

adamjstewart commented:
Yep, same issue, same fix. Check in `self.config_vars["base"] / lib` should solve the problem.

rountree commented:
Got a bit further but still hits an error.
```
==> Installing cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz
==> No binary for cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/f3/f316b40053466f9a416adf981efda41b160ca859e97f6a484b447ea299ff26aa.tar.gz
==> No patches needed for cmake
==> cmake: Executing phase: 'bootstrap'
==> cmake: Executing phase: 'build'
==> cmake: Executing phase: 'install'
==> cmake: Successfully installed cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz
  Fetch: 4.55s.  Build: 3m 24.31s.  Total: 3m 28.86s.
[+] /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz
==> Installing clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz
==> No binary for clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz found: installing from source
==> Already patched clingo-bootstrap
==> clingo-bootstrap: Executing phase: 'cmake'
==> Error: NoLibrariesError: Unable to locate python libraries in /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/lib

/Users/rountree4/repos/spack/var/spack/repos/builtin/packages/python/package.py:1098, in libs:
       1095
       1096        msg = "Unable to locate {} libraries in {}"
       1097        libdir = self.config_vars["LIBDIR"]
  >>   1098        raise spack.error.NoLibrariesError(msg.format(self.name, libdir))

See build log for details:
  /var/folders/0p/2_2pgh5x52z6pl6xxlt2cyv0001l2r/T/rountree4/spack-stage/spack-stage-clingo-bootstrap-spack-ltksuot66nuoxeb4bt2pulyohi34arxz/spack-build-out.txt

==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %apple-clang target=aarch64" due to the following failures:
github-actions-v0.1 raised ValueError: source is not trusted
spack-install raised InstallError: Terminating after first install failure: NoLibrariesError: Unable to locate python libraries in /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/lib
Run `spack --debug ...` for more detailed errors
     1028.17 real      1316.75 user       360.98 sys
```

adamjstewart commented:
Is that with eb36ad8b5f4e35fd043d460fa106d916adf63884 or 5c61bc7a4e419a0d1a5088d9dc07bb1a19536997? The former would explain that bug, but the latter should have fixed it.

rountree commented:
That's odd.... you're right, no idea how that happened.  Now definitely have the latter.  Retesting....

rountree commented:
This is almost certainly due to the cache.  Will remove that and start over.
```
1 error found in build log:
     9     -- Detecting C compile features - done
     10    -- Detecting CXX compiler ABI info
     11    -- Detecting CXX compiler ABI info - done
     12    -- Check for working CXX compiler: /Users/rountree4/repos/spack/lib/spack/env/clang/clang++ - skipped
     13    -- Detecting CXX compile features
     14    -- Detecting CXX compile features - done
  >> 15    CMake Error at /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.
           cmake:230 (message):
     16      Could NOT find Python (missing: Python_LIBRARIES Interpreter Development
     17      Development.Module Development.Embed) (found version "3.8.9")
     18    Call Stack (most recent call first):
     19      /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_F
           PHSA_FAILURE_MESSAGE)
     20      /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPython.cmake:561 (find_package_handle_s
           tandard_args)
     21      CMakeLists.txt:130 (find_package)
```

rountree commented:
Eh, not yet....
```
1 error found in build log:
  >> 3    CMake Error at /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.c
          make:230 (message):
     4      Could NOT find Python (missing: Python_LIBRARIES Interpreter Development
     5      Development.Module Development.Embed) (found version "3.8.9")
     6    Call Stack (most recent call first):
     7      /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FP
          HSA_FAILURE_MESSAGE)
     8      /Users/rountree4/repos/spack/local/user_cache/bootstrap/store/darwin-monterey-aarch64/apple-clang-13.1.6/cmake-3.23.2-movcuolmo7g2puqquibpis6j5bt253tz/share/cmake-3.23/Modules/FindPython.cmake:561 (find_package_handle_st
          andard_args)
     9      CMakeLists.txt:130 (find_package)
```

git log:
```
commit 15401ac94c2e899a3df8b72a4c4ec3165b2ca749 (HEAD -> develop)
Merge: 585623358c 5c61bc7a4e
Author: Rountree <rountree4@exegete.hsd1.ca.comcast.net>
Date:   Wed Aug 3 19:33:45 2022 -0700

    Merge remote-tracking branch 'adam/fixes/python' into develop

commit 5c61bc7a4e419a0d1a5088d9dc07bb1a19536997 (adam/fixes/python)
Author: Adam J. Stewart <ajstewart426@gmail.com>
Date:   Wed Aug 3 18:57:14 2022 -0700

    Fix libs too

commit 585623358c3e6b0963ed1b3fa925b6c71b0d9003
Merge: 9a48035e49 eb36ad8b5f
Author: Rountree <rountree4@exegete.hsd1.ca.comcast.net>
Date:   Wed Aug 3 18:54:13 2022 -0700

    Merge remote-tracking branch 'adam/fixes/python' into develop

commit eb36ad8b5f4e35fd043d460fa106d916adf63884
Author: Adam J. Stewart <ajstewart426@gmail.com>
Date:   Wed Aug 3 18:46:26 2022 -0700

    Explicitly define Python library to use

commit 9a48035e49491875b192ce104d1e3f6a13d570ae (origin/develop, origin/HEAD)
Author: Massimiliano Culpo <massimiliano.culpo@gmail.com>
Date:   Mon Jun 20 17:49:33 2022 +0200

    asp: refactor low level API to permit the injection of configuration
    
    This allows writing extension commands that can benchmark
    different configurations in clingo, or try different
    configurations for a single test.
```

adamjstewart commented:
Hmm, so it finds the lib, but CMake doesn't like it. I assume it's looking for the full path to the library name, not a directory/file/libname. But you can try switching `libs[0]` to `libs.directories[0]` or `libs.names[0]`? Idk, not clear why clingo CMake isn't happy. Maybe @tgamblin @alalazo can help.

rountree commented:
Would it be useful to do the full Xcode install and rerun, or is that a known good build?

The M1 has been out for a while now and it looks like I'm the first person to get this to fail by not having the full Xcode install.  We could just say that's a requirement and call it day.

(I'd like to get this fixed on general principles, but this is becoming a bit of a time sink for you.)

adamjstewart commented:
Nah, I also want to get it working. The full Xcode suite is known to work (that's what I'm using on my M1 laptop). It shouldn't be impossible to get the CLT-only environment working.

alalazo commented:
> We could just say that's a requirement and call it day.

@rountree It currently is a requirement. In the [Getting Started](https://spack.readthedocs.io/en/latest/getting_started.html) page we say: 

> on macOS, XCode is required.

If this PR gets merged we can relax the requirement.

adamjstewart commented:
@alalazo previously, only Xcode command line tools were required. Packages like Qt require the full suite, but the rest of Spack should only require the CLT. It seems like bootstrapping fails if only CLT is available.

rountree commented:
@adamjstewart This has me blocked using spack on my new laptop, so in case you're planning on spending some more time on it this week, I'm going to go ahead and install the full Xcode and call it a day.

tgamblin commented:
Ok I fixed this.  The last commit has the details.  The issues were:

The Python that comes with XCode command line tools is a bit weird. This fixes two
issues that were preventing it from bootstrapping:

- [x] CommandLineTools python does not come with a `python-config` executable. If you don't have one of these, CMake's FindPython will, for some reason, assume that you want Python 2, so you need to set `CLINGO_PYTHON_VERSION` to ensure that clingo tells `find_package(Python <VERSION>)` the right thing.

- [x] We weren't setting `PYTHONHOME` correctly in Python's `package.py`. We were setting it to the `prefix` from Python `sysconfig`, but `prefix` tells you where the executable lives. CommandLineTools python reports its prefix as:

    `/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8`
    
    but that doesn't exist.  It looks like Apple builds the full python and just copies pieces of it into command line tools. `PYTHONHOME` is supposed to tell you where the default python *library* location is. So you have to look at the `base`, which `sysconfig` reports as
    
    `/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8`
    
    On most systems this is probably the same as `prefix`, but not with CommandLineTools, which has its system library in a different place.

tgamblin commented:
Can others verify that:
1. It works for them on M1 with Monterey and CommandLineTools?
2. It works with full Xcode?

rountree commented:
@tgamblin 

https://github.com/spack/spack/commit/1d64caf37fbdae30e09d3d584e99aa6c19314ecb gives me the missing OpenSSL problem.  New to me, but I think @trws mentioned it earlier.
Same error bootstrapping on the development branch.

Running with full Xcode install.  No other tools installed.

<details><summary>Details</summary>
```
2 errors found in build log:
     357    -- Looking for connect in socket;-framework SystemConfiguration - not found
     358    -- Looking for gethostbyname in c
     359    -- Looking for gethostbyname in c - found
     360    -- Looking for gethostname
     361    -- Looking for gethostname - found
     362    -- Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the system variable OPENSSL_ROOT_DIR (missing: OPENS
            SL_CRYPTO_LIBRARY OPENSSL_INCLUDE_DIR)
  >> 363    CMake Error at Utilities/cmcurl/CMakeLists.txt:586 (message):
     364      Could not find OpenSSL.  Install an OpenSSL development package or
     365      configure CMake with -DCMAKE_USE_OPENSSL=OFF to build without OpenSSL.
     366    
     367    
     368    -- Configuring incomplete, errors occurred!
     369    See also "/private/var/folders/0p/2_2pgh5x52z6pl6xxlt2cyv0001l2r/T/rountree4/spack-stage/spack-stage-cmake-3.23.3-rackjzvlzgv
            mv4mv32su4aoviasxgcss/spack-src/CMakeFiles/CMakeOutput.log".
     370    See also "/private/var/folders/0p/2_2pgh5x52z6pl6xxlt2cyv0001l2r/T/rountree4/spack-stage/spack-stage-cmake-3.23.3-rackjzvlzgv
            mv4mv32su4aoviasxgcss/spack-src/CMakeFiles/CMakeError.log".
     371    ---------------------------------------------
  >> 372    Error when bootstrapping CMake:
     373    Problem while running initial CMake
     374    ---------------------------------------------
```
</details>

rountree commented:
Ah, 40beae7491 (Adam's patch) is giving the same error and that was working before.

adamjstewart commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1100,20 +1106,33 @@ def libs(self):
+        # This varies by system, especially on macOS where the command line tools are
+        # installed in a very different directory from the system python interpreter.
+        py_version = str(self.version.up_to(2))
+        candidates = list(
```
Idk if we need to convert it back to a list here

tgamblin replied:
I guess we can omit it -- do you want me to?

adamjstewart replied:
Yeah, if it isn't needed, might as well remove it to keep the code simpler

tgamblin replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1100,20 +1106,33 @@ def libs(self):
+                )
+            )
 
         headers.directories = [os.path.dirname(config_h)]
```
Is this correct for all layouts? This was intended for layouts like:
```
<prefix>/include/pythonX.Y/pyconfig.h
```
where a user would use:
```c
#include <pythonX.Y/pyconfig>
```
But if the location is something like:
```
<prefix>/Headers/pyconfig.h
```
what would the proper directory be?

tgamblin replied:
It doesn't really matter as long as you find the python includes.  On macOS, `include/python3.8` is a symlink to `Headers` in the framework directory, so we would find the `include/python3.8` option first, given the order I listed things.

I'm not sure what the "proper" directory is, but I prefer to find the ones that looks more like a standard unix layout first, because it's more likely that code that uses `..` to infer things from discovered paths will be correct.  e.g., I think most code in a generic build assumes that python headers will be in `include/pythonX.Y` and that `bin`, `lib`, etc. are in `../..` from there.  If you return the `Headers` directory, that breaks because it's only one deep (because Apple or something -- I guess it's b/c that is how their frameworks work -- you pick the python you want, THEN include its `Headers` directory.

adamjstewart replied:
Maybe we should change this to:
```suggestion
        if not headers.directories[0].endswith('Headers'):
            headers.directories = [os.path.dirname(config_h)]
```

tgamblin replied:
oh I see what you're saying -- nice catch.  Yeah lemme see if I can either do that or simplify it further.

adamjstewart replied:
Nvm, this is correct as is.

adamjstewart commented:
@rountree I think the OpenSSL problem is different, and I'm unfortunately unsure how to reproduce or fix it.

adamjstewart commented:
Confirmed that this works with the full Xcode suite on macOS 12.5.

adamjstewart commented:
Also confirmed that bootstrapping still works with Spack-installed Python.

tgamblin committed:
clingo: fix bootstrapping on M1/monterey with command line tools

The Python that comes with XCode command line tools is a bit weird.

This fixes two issues that were preventing it from bootstrapping:

- [x] CommandLineTools python does not come with a `python-config` executable. If you
      don't have one of these, CMake's FindPython will, for some reason, assume that you
      want Python 2, so you need to set `CLINGO_PYTHON_VERSION` to ensure that clingo
      tells `find_package(Python <VERSION>)` the right thing.

- [x] We weren't setting `PYTHONHOME` correctly in Python's `package.py`. We were
      setting it to the `prefix` from Python `sysconfig`, but `prefix` tells you where
      the executable lives. CommandLineTools python reports its prefix as

          /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8

      but that doesn't exist. It looks like Apple builds the full python and just copies
      pieces of it into command line tools. PYTHONHOME is supposed to tell you where the
      default python library location is. So you have to look at the `base`, which
      `sysconfig` reports as

          /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8

      On most systems this is probably the same as `prefix`, but not with
      CommandLineTools, which has its system library in a different place.

The second change here was cherry-picked to 0d981a012d before merging and doesn't show
up here.

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

adamjstewart commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1061,59 +1061,77 @@ def find_library(self, library):
-        raise spack.error.NoLibrariesError(msg.format(self.name, libdir))
+        raise spack.error.NoLibrariesError(
+            "Unable to locate {} libraries in any of these locations:".format(self.name),
+            "\n".join(candidates),
```
`candidates` is a list of library names, not locations.

tgamblin replied:
fixed

rountree commented:
@adamjstewart @tgamblin `spack clean --all --bootstrap` did the trick.  I can also confirm things are working with a full Xcode install.

tgamblin commented:
Only thing failing in the pipeline here is the SYCL issue on `develop` that was fixed in #32186.
